runtime/interpreter/value.go

```diff
@@ -1688,6 +1703,44 @@ func (v *StringValue) ForEach(
 	}
 }
 
+func (v *StringValue) Contains(inter *Interpreter, other *StringValue) BoolValue {
+
+	// Meter computation as if the string was iterated.
+	// This is a conservative over-estimation.
+	inter.ReportComputation(common.ComputationKindLoop, uint(len(v.Str)*len(other.Str)))
+
+	v.prepareGraphemes()
+
+	for {
+		start, _ := v.graphemes.Positions()
+		remainder := v.Str[start:]
+		if strings.HasPrefix(remainder, other.Str) {
+
+			// Check the end is a grapheme cluster boundary
+			expectedEnd := start + len(other.Str)
+			for {
+				_, end := v.graphemes.Positions()
+				if end == expectedEnd {
+					return TrueValue
+				} else if end > expectedEnd {
+					return FalseValue
+				}
+
+				if !v.graphemes.Next() {
+					break
+				}
+			}
```

Comment:
Asking out of curiosity, why would this inner loop be needed? Do you know any example?

Reply:
Great question! The tests contain an example: in ` \u{1F476}\u{1F3FB}`, ` \u{1F476}` is NOT a substring -- though the boundary start matches, the end does not.

BTW, I optimized the function in [1fcafe0](https://github.com/onflow/cadence/pull/3455/commits/1fcafe0f7bdba56e719f7a5dfd70a90cb3fcb165), but it still contains the grapheme cluster boundary end check.

Reply:
oh  see, thank you!